crates/sui-core/src/authority.rs

```diff
@@ -1590,100 +1592,173 @@ impl AuthorityState {
+        // Payment might be empty here, but it's fine we'll have to deal with it later after reading all the input objects.
+        let payment = gas_objects.unwrap_or_default();
+        let expiration = match epoch {
+            None => TransactionExpiration::None,
+            Some(epoch) => TransactionExpiration::Epoch(epoch),
         };
-        let gas_status =
-            SuiGasStatus::new(max_tx_gas, gas_price, reference_gas_price, protocol_config)?;
-
-        let gas_object_id = ObjectID::random();
-        // give the gas object 2x the max gas to have coin balance to play with during execution
-        let gas_object = Object::new_move(
-            MoveObject::new_gas_coin(SequenceNumber::new(), gas_object_id, max_tx_gas * 2),
-            Owner::AddressOwner(sender),
-            TransactionDigest::genesis_marker(),
-        );
+        let transaction = TransactionData::V1(TransactionDataV1 {
+            kind: transaction_kind.clone(),
+            sender,
+            gas_data: GasData {
+                payment,
+                owner,
+                price: gas_price,
+                budget: gas_budget,
+            },
+            expiration,
+        });
 
-        let input_object_kinds = transaction_kind.input_objects()?;
-        let receiving_object_refs = transaction_kind.receiving_objects();
-        let (input_objects, receiving_objects) = self
+        transaction.check_version_supported(protocol_config)?;
+        transaction.validity_check_no_gas_check(protocol_config)?;
+
+        let input_object_kinds = transaction.input_objects()?;
+        let receiving_object_refs = transaction.receiving_objects();
+
+        if !skip_checks {
+            sui_transaction_checks::deny::check_transaction_for_signing(
+                &transaction,
+                &[],
+                &input_object_kinds,
+                &receiving_object_refs,
+                &self.transaction_deny_config,
+                &self.database,
+            )?;
+        }
+
+        let (mut input_objects, receiving_objects) = self
             .input_loader
             .read_objects_for_dev_inspect(
                 &input_object_kinds,
                 &receiving_object_refs,
                 protocol_config,
             )
             .await?;
-        let (gas_object_ref, checked_input_objects) =
-            sui_transaction_checks::check_dev_inspect_input(
+
+        // Create and use a dummy gas object if there is no gas object provided.
+        let dummy_gas_object = Object::new_gas_with_balance_and_owner_for_testing(
+            DEV_INSPECT_GAS_COIN_VALUE,
+            transaction.gas_owner(),
+        );
+
+        let gas_objects = if transaction.gas().is_empty() {
+            let gas_object_ref = dummy_gas_object.compute_object_reference();
+            vec![gas_object_ref]
+        } else {
+            transaction.gas().to_vec()
+        };
+
+        let (gas_status, checked_input_objects) = if skip_checks {
```

Comment:
There seems to be some interesting nuances between the interfaces for these various kinds of check (e.g. some of these APIs seem to expect the gas object in the inputs, some not, etc etc).  It's probably something we can simplify (later), but for now, can you add some comments to this part to call out those differences?

Reply:
Yup, this is precisely why I did not collapse the `else` and `if` - not collapsing them would make the conditions more clear and code more readable imo. I'll add some comments explaining this.